runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/jobsubmission/JobService.java

```diff
@@ -0,0 +1,309 @@
+      ArtifactStagingServiceProvider artifactStagingServiceProvider, JobInvoker invoker) {
+    this.artifactStagingServiceProvider = artifactStagingServiceProvider;
+    this.invoker = invoker;
+
+    this.preparations = new ConcurrentHashMap<>();
+    this.invocations = new ConcurrentHashMap<>();
+  }
+
+  @Override
+  public void prepare(
+      PrepareJobRequest request,
+      StreamObserver<PrepareJobResponse> responseObserver) {
+    try {
+      LOG.trace("{} {}", PrepareJobRequest.class.getSimpleName(), request);
+      // insert preparation
+      String preparationId =
+          String.format("%s_%d", request.getJobName(), ThreadLocalRandom.current().nextInt());
+      GrpcFnServer<ArtifactStagingService> stagingService =
+          artifactStagingServiceProvider.forJob(preparationId);
+      Struct pipelineOptions = request.getPipelineOptions();
+      if (pipelineOptions == null) {
+        LOG.trace("PIPELINE OPTIONS IS NULL");
+        throw new NullPointerException("Encountered null pipeline options.");
+            /*
+        LOG.debug("Encountered null pipeline options.  Using default.");
+        pipelineOptions = Struct.getDefaultInstance();
+        */
+      } else {
+        LOG.trace("PIPELINE OPTIONS IS NOT NULL");
+      }
+      LOG.trace("PIPELINE OPTIONS {} {}", pipelineOptions.getClass(), pipelineOptions);
+      JobPreparation preparation =
+          JobPreparation
+              .builder()
+              .setId(preparationId)
+              .setPipeline(request.getPipeline())
+              .setOptions(pipelineOptions)
+              .setStagingService(stagingService)
+              .build();
+      JobPreparation previous = preparations.putIfAbsent(preparationId, preparation);
+      if (previous != null) {
+        String errMessage =
+            String.format("A job with the preparation ID \"%s\" already exists.", preparationId);
+        StatusException exception = Status.NOT_FOUND.withDescription(errMessage).asException();
+        responseObserver.onError(exception);
+        return;
+      }
+
+      // send response
+      PrepareJobResponse response =
+          PrepareJobResponse
+              .newBuilder()
+              .setPreparationId(preparationId)
+              .setArtifactStagingEndpoint(stagingService.getApiServiceDescriptor())
+              .build();
+      responseObserver.onNext(response);
+      responseObserver.onCompleted();
+    } catch (Exception e) {
+      LOG.error("Could not prepare job with name {}", request.getJobName(), e);
+      responseObserver.onError(Status.INTERNAL.withCause(e).asException());
+    }
+  }
+
+  @Override
+  public void run(
+      RunJobRequest request, StreamObserver<RunJobResponse> responseObserver) {
+    try {
+      LOG.trace("{} {}", RunJobRequest.class.getSimpleName(), request);
+
+      // retrieve job preparation
+      String preparationId = request.getPreparationId();
+      JobPreparation preparation = preparations.get(preparationId);
```

Comment:
We should mark preparationId as used and should not reuse it.

Reply:
Why not?  I thought that a plausible use case was to prepare a job once and then run it multiple times.

Reply:
Re-using preparation id will be very tricky especially when we have not assigned a life span to the preparations it.
Also there is this question of re-trying a preparation id if the 1st call for run fails. In this case starting a new job for the same preparation id is not the right intent.
We can certainly reuse the preparation id but then we should clearly outline the life span of it and provide clear ways to reuse it.
To keep things simple I would recommend simply not reusing the preparation id.